### PR TITLE
Collision revamp and necessary bug fixes to enable playtesting of collision revamp

### PR DESCRIPTION
**Pull request by ministerofinformation**
Thank you for submitting a pull request and becoming a contributor to the Vega Strike Core Engine.

Please answer the following:

Code Changes:
- [X] Have the PR Validation Tests been run? See https://github.com/vegastrike/Vega-Strike-Engine-Source/wiki/Pull-Request-Validation
- [ ] This is a documentation change only

Issues:
#415
#502  
#498
#494 
#487 
#486 

Purpose:
- What is this pull request trying to do? 1) Fix errors in collision response code (velocity != force) 2) re-add torque to collision response after removal during refactoring 3) fix compile/crash/critical errors that prevented playtesting after merging in recent PRs.
- What release is this for?
- Is there a project or milestone we should apply this to?

### CODE REVIEW

**Comment by ministerofinformation** [comment]
Known caveats:

I expect to make some changes on review - at the very least, the fix for the global config is ... a duct tape level workaround dropped in so I could test, commit, and go to sleep before it gets light out :-P. Let's figure out what we want to do to have it be clean - but I wanted to at least have a working Windows build option again.

Likewise, the #ifdef for M_PI -- if we want to be pushing the defines to CMAKE, that's fine too.

The collision code has some validation logic that should eventually be factored out to testing code, and it can be made leaner and cleaner.

The collision code changes were designed to be localized to collision.cpp/h - as such, there are some both fundamental and dataset issues that have not been resolved with respect to moment of inertia, as well as invariants that would be better enforced in other portions of the code (e.g. getMass() should always return a non-zero value, either by ensuring that minimum mass is specified at initialization or by hiding the std::max operation behind the getter, etc.) These represent some more systemic changes to enhance cleanliness, but they'll have to wait for another PR. I'll try to be sure to file issues on all of the things that came up during this feature redevelopment.

File: engine/src/main.cpp
Comment (stephengtuggy) [inline engine/src/main.cpp:388]: No `memcpy`. Bad `memcpy`. Not for complex types like structs and classes, at any rate.

The = operator sounds like a better idea.

File: engine/src/configuration/configuration.cpp
Comment (stephengtuggy) [inline engine/src/configuration/configuration.cpp:4]: Nice

File: engine/src/cmd/collision.cpp
Comment (stephengtuggy) [inline engine/src/cmd/collision.cpp:338]: Wow. My head is spinning right about now. Nice job, @ministerofinformation

**Review by stephengtuggy — CHANGES_REQUESTED** [review]
I would like to request that the `memcpy` be replaced with something else before we merge this. Otherwise, looks good!

File: engine/src/main.cpp
Comment (ministerofinformation) [reply engine/src/main.cpp:388]: Yup. Quick and dirty 3am code is ... exactly-as-awful-as-expected 3am code. I don't actually _want_ this accepted before I fix this, but I also don't want to delay the rest of this PR being reviewed/tested any further. I expect to be on baby duty for much of the day, so I'll try to get to this to tonight/tomorrow. This is known-flawed-and-dangerous for non-POD structs, but it enables testing of the config functionality, etc. 

Re = operator -- that will work for now and be generally potentially useful indefinitely. Longer term, I think the discussion was moving towards upgrading Configuration to have methods, while keeping its members as simple structs, and (re)init/setters would then be methods in Configuration. If we want it to initialize at construction, Configuration can only init to defaults, and we should be explicit in documenting this in the code and using obvious patterns (i.e. If construction is default, just use the default, not the call that looks like it's getting a value, and to keep default specification singular, the XML reader function used to init should then not have a default field at all)

File: engine/src/cmd/collision.cpp
Comment (ministerofinformation) [reply engine/src/cmd/collision.cpp:338]: If your head is spinning, are there more comments I could add to make it more self-contained without dumping the whole Wikipedia article contents in (the figure is actually quite useful for building an intuition of what's going on)? I'd like this to be as clearly documented as possible, so as to separate the math of the collision from future renovations on the moment of inertia implementation and various getters/non-local invariants we want to adjust/enforce (applying Newtonian physics to zero-mass objects tends to end poorly in more than the collision code, etc.)

Comment (stephengtuggy) [reply engine/src/cmd/collision.cpp:338]: No, it's fine, thanks. Just a lot of code to read through. :-)

Comment (Loki1950) [reply engine/src/cmd/collision.cpp:338]: Just standard vector arithmetic @stephengtuggy

**Comment by stephengtuggy** [comment]
I ran a bunch more of the play tests today. Looking good so far. Ships actually bounce and spin away now 💯

**Comment by ministerofinformation** [comment]
> I ran a bunch more of the play tests today. Looking good so far. Ships actually bounce and spin away now 💯

If you have the time, I'd also recommend testing with some different values for the elasticity constant "inelastic_scale" (for historical reasons, 1-elasticity) in the physics section of the config xml to make sure you're seeing the expected differences, to test the #502 aspect of the PR.

**Comment by ministerofinformation** [comment]
After sleeping on it, I went with adding a new member function versus putting in an = operator. I think this more clearly embeds the pattern of what we want to happen at each point in time: A specific default value at construction time (everything is initialized to some valid value), followed by an adjustment to said values after the user configuration file has been loaded (existing value used as default in the GameConfig::GetVariable function, so that there aren't two copies of the default value in play). We may want to consider cleaning this up further by writing a companion to GameConfig::GetVariable, say, GameConfig::OverrideVariableIfSpecified, or somesuch, that takes a mutable reference to the existing value and only modifies it if there's an overriding value. The current pattern is clear enough, and flexible enough, but a little more verbose.

**Comment by stephengtuggy** [comment]
@ministerofinformation I like the latest change a lot. Makes things much clearer. I will plan on doing some more playtesting here shortly.

File: engine/src/cmd/collision.cpp
Comment (BenjamenMeyer) [inline engine/src/cmd/collision.cpp:152]: The `0.5` should be a constant declared somewhere; magic values shouldn't appear as anything else.

File: engine/src/cmd/collision.cpp
Comment (BenjamenMeyer) [inline engine/src/cmd/collision.cpp:175]: same with the `2`

File: engine/src/cmd/collision.cpp
Comment (BenjamenMeyer) [inline engine/src/cmd/collision.cpp:212]: same for the 0.667

Comment (BenjamenMeyer) [reply engine/src/cmd/collision.cpp:212]: possibly the `1` as well

File: engine/src/cmd/collision.cpp
Comment (BenjamenMeyer) [inline engine/src/cmd/collision.cpp:235]: would be nice if the fuzzy compare was abstracted to its own inline function for re-use

**Review by BenjamenMeyer — APPROVED** [review]
would be nice to move the `magic numbers` to constants so we can understand their meaning better; but as there's a lot of them in the code right now I'm not going to block.
Also would be nice to abstract the fuzzy float/double comparison, but again, not blocking.

File: engine/src/cmd/collision.cpp
Comment (ministerofinformation) [reply engine/src/cmd/collision.cpp:212]: The 1 is just an inversion - I'd leave the 1 as is.

File: engine/src/cmd/collision.cpp
Comment (ministerofinformation) [reply engine/src/cmd/collision.cpp:175]: Agreed - albeit in this specific function everything is currently magic, so it'll just be moved to a local var just to not be nameless, rather than something more configurable.

File: engine/src/cmd/collision.cpp
Comment (ministerofinformation) [reply engine/src/cmd/collision.cpp:212]: > same for the 0.667

Agreed -  the 0.667, as well as the need to access radial_size is a bandaid over a longer-term fix; it will eventually go away when we renovate the moment of inertia to not be scalar in the first place and also fix the dataset accordingly to specify the proper 3x3 matrices.

File: engine/src/cmd/collision.cpp
Comment (ministerofinformation) [reply engine/src/cmd/collision.cpp:235]: Probably a good idea to do so later, make it an inline function with both a percent and absolute bound - where in the codebase do you think is the best place to have testing/logging/etc. adjacent functions specified?

File: engine/src/cmd/collision.cpp
Comment (ministerofinformation) [reply engine/src/cmd/collision.cpp:152]: Agreed - should factor this one out into a something representing the fraction of deltaKE assigned to this unit vs. the other unit.

File: engine/src/cmd/collision.cpp
Comment (BenjamenMeyer) [reply engine/src/cmd/collision.cpp:235]: Not sure; right now, it could just site in a function above this one.

File: engine/src/cmd/collision.cpp
Comment (BenjamenMeyer) [reply engine/src/cmd/collision.cpp:175]: doesn't have to be configurable - the point of not having magic numbers is that there's no meaning in the number; using a variable (configurable) or a constant (non-configurable) value instead conveys the meaning of the value.
For now, even a simple constant right before usage is good. If we find a lot of the same, then it should be moved to a file header scope as appropriate.

**Review by stephengtuggy — APPROVED** [review]
Looks good to me! Sorry, I didn't realize I was holding up this PR.

Can we go ahead and merge this? I think it's urgent that we get master fixed.

**Comment by stephengtuggy** [comment]
I have finished play testing this code to my satisfaction.

**Pull request merged**
